### PR TITLE
YD-512 Analysis engine service testability

### PR DESCRIPTION
**Pull request by aristotelos**
Tried to take a first step to increase testability.

- Separated `ActivityUpdateService` from `AnalysisEngineService`
- Split unit tests in two
- Made `UserAnonymizedEntityHolder` static
- Fixed a bug that `ensureMinimumDurationOneMinute` of 0:59 returned 1:59 instead of 1:00
- Added `equals` to `ActivityPayload` to be able to assert on it

### CODE REVIEW

File: analysisservice/src/test/java/nu/yona/server/analysis/service/ActivityUpdateServiceTest.java [hunk 1]
```diff
@@ -0,0 +1,743 @@
...
+		ZonedDateTime t1 = now();
+
+		service.addActivity(userAnonymizedEntityHolder, createPayload(t1, t1), GoalDto.createInstance(socialGoal),
+				Optional.empty());
+
+		verifyNoGoalConflictMessagesCreated();
+	}
+
+	@Test
+	public void addActivity_afterConflictIntervalLastRegisteredActivity_newGoalConflictMessageCreated()
```
Comment (Bert-R) [inline analysisservice/src/test/java/nu/yona/server/analysis/service/ActivityUpdateServiceTest.java]: I wouldn't add 'new' in the name. Anything you create is new, right? :smiley:

File: analysisservice/src/test/java/nu/yona/server/analysis/service/ActivityUpdateServiceTest.java [hunk 2]
```diff
@@ -0,0 +1,743 @@
...
+		ActivityDto lastRegisteredActivity = ActivityDto.createInstance(existingActivityEntity);
+
+		service.updateTimeLastActivity(mockUserAnonymizedEntityHolder, createPayload(t1, t1),
+				GoalDto.createInstance(gamblingGoal), lastRegisteredActivity);
+
+		verify(mockUserAnonymizedEntityHolder, atLeastOnce()).getEntity();
+	}
+
+	@Test
+	public void updateTimeLastActivity_startTimeBefore_activityStartTimeUpdated()
```
Comment (Bert-R) [inline analysisservice/src/test/java/nu/yona/server/analysis/service/ActivityUpdateServiceTest.java]: What about ``startTimeEarlier``?

File: analysisservice/src/test/java/nu/yona/server/analysis/service/ActivityUpdateServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,743 @@
...
+
+		service.updateTimeLastActivity(userAnonymizedEntityHolder, createPayload(t1, t2), GoalDto.createInstance(gamblingGoal),
+				lastRegisteredActivity);
+
+		assertThat("Expect start time updated", existingActivityEntity.getStartTimeAsZonedDateTime(), equalTo(t1));
+		assertThat("Expect end time same", existingActivityEntity.getEndTimeAsZonedDateTime(), equalTo(t3));
+	}
+
+	@Test
+	public void updateTimeLastActivity_endTimeAfter_activityEndTimeUpdated()
```
Comment (Bert-R) [inline analysisservice/src/test/java/nu/yona/server/analysis/service/ActivityUpdateServiceTest.java]: What about ``endTimeLater``?

File: analysisservice/src/test/java/nu/yona/server/analysis/service/ActivityUpdateServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,743 @@
...
+		DayActivity existingDayActivityEntity = mockExistingActivity(gamblingGoal, t1, t1, "Lotto");
+		Activity existingActivityEntity = existingDayActivityEntity.getLastActivity(deviceAnonId);
+
+		service.updateTimeExistingActivity(mockUserAnonymizedEntityHolder, createPayload(t1, t1), existingActivityEntity);
+
+		verify(mockUserAnonymizedEntityHolder, atLeastOnce()).getEntity();
+	}
+
+	@Test
+	public void updateTimeExistingActivity_startTimeBefore_activityStartTimeUpdated()
```
Comment (Bert-R) [inline analysisservice/src/test/java/nu/yona/server/analysis/service/ActivityUpdateServiceTest.java]: What about ``startTimeEarlier``?

File: analysisservice/src/test/java/nu/yona/server/analysis/service/ActivityUpdateServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,743 @@
...
+		Activity existingActivityEntity = existingDayActivityEntity.getLastActivity(deviceAnonId);
+
+		service.updateTimeExistingActivity(userAnonymizedEntityHolder, createPayload(t1, t2), existingActivityEntity);
+
+		assertThat("Expect start time updated", existingActivityEntity.getStartTimeAsZonedDateTime(), equalTo(t1));
+		assertThat("Expect end time same", existingActivityEntity.getEndTimeAsZonedDateTime(), equalTo(t3));
+	}
+
+	@Test
+	public void updateTimeExistingActivity_endTimeAfter_activityEndTimeUpdated()
```
Comment (Bert-R) [inline analysisservice/src/test/java/nu/yona/server/analysis/service/ActivityUpdateServiceTest.java]: What about ``endTimeLater``?

File: analysisservice/src/main/java/nu/yona/server/analysis/service/ActivityUpdateService.java [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+		Activity activity = dayActivity.getLastActivity(payload.deviceAnonymized.getId());
+		if (payload.startTime.isBefore(lastRegisteredActivity.getStartTime()))
+		{
+			activity.setStartTime(payload.startTime.toLocalDateTime());
+		}
+		if (payload.endTime.isAfter(lastRegisteredActivity.getEndTime()))
+		{
+			activity.setEndTime(payload.endTime.toLocalDateTime());
+		}
+		cacheService.updateLastActivityForUser(payload.userAnonymized.getId(), payload.deviceAnonymized.getId(),
```
[the remark this answers is not in the document]
Comment (aristotelos) [reply analysisservice/src/main/java/nu/yona/server/analysis/service/ActivityUpdateService.java]: As this method updates the last activity, `shouldUpdateCache` always returns `true`. So therefore I simplified the logic.

[the remark this answers is not in the document]
Comment (Bert-R) [reply analysisservice/src/main/java/nu/yona/server/analysis/service/ActivityUpdateService.java]: OK, clear.

File: analysisservice/src/main/java/nu/yona/server/analysis/service/ActivityUpdateService.java [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+import org.springframework.stereotype.Service;
+
+import nu.yona.server.analysis.entities.Activity;
+import nu.yona.server.analysis.entities.ActivityRepository;
+import nu.yona.server.analysis.entities.DayActivity;
+import nu.yona.server.analysis.entities.DayActivityRepository;
+import nu.yona.server.analysis.entities.GoalConflictMessage;
+import nu.yona.server.analysis.entities.WeekActivity;
+import nu.yona.server.analysis.entities.WeekActivityRepository;
+import nu.yona.server.analysis.service.AnalysisEngineService.ActivityPayload;
```
[the remark this answers is not in the document]
Comment (aristotelos) [reply analysisservice/src/main/java/nu/yona/server/analysis/service/ActivityUpdateService.java]: Agreed

File: analysisservice/src/main/java/nu/yona/server/analysis/service/AnalysisEngineService.java [hunk 8]
```diff
@@ -620,6 +434,28 @@ static ActivityPayload createInstance(UserAnonymizedDto userAnonymized, DeviceAn
 					startTime.withZoneSameInstant(userTimeZone), endTime.withZoneSameInstant(userTimeZone),
 					Optional.of(application));
 		}
+
+		@Override
+		public boolean equals(Object other)
+		{
+			if (other == null)
```
[the remark this answers is not in the document]
Comment (aristotelos) [reply analysisservice/src/main/java/nu/yona/server/analysis/service/AnalysisEngineService.java]: OK, will update

File: analysisservice/src/test/java/nu/yona/server/analysis/service/ActivityUpdateServiceTest.java (hunk 1 above)
Comment (aristotelos) [reply analysisservice/src/test/java/nu/yona/server/analysis/service/ActivityUpdateServiceTest.java]: OK, will change

File: analysisservice/src/test/java/nu/yona/server/analysis/service/ActivityUpdateServiceTest.java (hunk 3 above)
Comment (aristotelos) [reply analysisservice/src/test/java/nu/yona/server/analysis/service/ActivityUpdateServiceTest.java]: Agreed

File: analysisservice/src/test/java/nu/yona/server/analysis/service/ActivityUpdateServiceTest.java [hunk 9]
```diff
@@ -0,0 +1,743 @@
...
+
+@RunWith(MockitoJUnitRunner.class)
+public class ActivityUpdateServiceTest
+{
+	private final Map<String, Goal> goalMap = new HashMap<>();
+
+	@Mock
+	private ActivityCategoryService mockActivityCategoryService;
+	@Mock
+	private ActivityCategoryService.FilterService mockActivityCategoryFilterService;
```
[the remark this answers is not in the document]
Comment (aristotelos) [reply analysisservice/src/test/java/nu/yona/server/analysis/service/ActivityUpdateServiceTest.java]: Removed

File: analysisservice/src/test/java/nu/yona/server/analysis/service/ActivityUpdateServiceTest.java [hunk 10]
```diff
@@ -0,0 +1,743 @@
...
+	@Mock
+	private ActivityRepository mockActivityRepository;
+	@Mock
+	private DayActivityRepository mockDayActivityRepository;
+	@Mock
+	private WeekActivityRepository mockWeekActivityRepository;
+	@Mock
+	private LockPool<UUID> userAnonymizedSynchronizer;
+	@Mock
+	private TransactionHelper transactionHelper;
```
[the remark this answers is not in the document]
Comment (aristotelos) [reply analysisservice/src/test/java/nu/yona/server/analysis/service/ActivityUpdateServiceTest.java]: Removed, sorry for the copy-paste

File: analysisservice/src/test/java/nu/yona/server/analysis/service/ActivityUpdateServiceTest.java [hunk 11]
```diff
@@ -0,0 +1,743 @@
...
+	@Mock
+	private DeviceService mockDeviceService;
```
[the remark this answers is not in the document]
Comment (aristotelos) [reply analysisservice/src/test/java/nu/yona/server/analysis/service/ActivityUpdateServiceTest.java]: Removed, sorry for the copy-paste

File: analysisservice/src/main/java/nu/yona/server/analysis/service/ActivityUpdateService.java [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+	@Autowired(required = false)
+	private ActivityRepository activityRepository;
+	@Autowired(required = false)
+	private DayActivityRepository dayActivityRepository;
+	@Autowired(required = false)
+	private WeekActivityRepository weekActivityRepository;
+	@Autowired
+	private ActivityCacheService cacheService;
+
+	public void updateLastMonitoredActivityDateIfRelevant(UserAnonymizedEntityHolder userAnonymizedHolder,
```
[the remark this answers is not in the document]
Comment (aristotelos) [reply analysisservice/src/main/java/nu/yona/server/analysis/service/ActivityUpdateService.java]: That's another great simplification! Thanks for suggesting.

**Comment by aristotelos** [comment]
Updated!

**Comment by Bert-R** [comment]
That's a major improvement!
We're currently facing a deployment issue on the Mobiquity test server. Once that's resolved and that build is deployed, I'll merge this build and push it to production.

**Pull request merged**